penguin/engine.py

```
@@ -1428,6 +1431,33 @@ async def _iteration_loop(
                         f"Preview: {last_response[:100]}..."
                     )
 
+                if (
+                    last_response
+                    and config.completion_phrases
+                    and any(phrase in last_response for phrase in config.completion_phrases)
+                    and not termination_detected
+                ):
+                    logger.info(
+                        f"Completion phrase detected in {config.mode} loop output without explicit action."
```

Ruff (0.15.10)

[warning] Logging statement uses f-string

(G004)